### PR TITLE
Job to Remove Datastore Tables for Unsupported Resource Types

### PR DESCRIPTION
**Pull request by JVickery-TBS**
feat(dev): implemented IDomainObjectModification;

- Implemented `IDomainObjectModification` instead of `IResourceUrlChange`.
- Queued new job callback for unsupported datastore resources.
- Added new config option `ckanext.xloader.limit_datastore_tables`.

See https://github.com/open-data/ckanext-canada/pull/1382 for more discussion.

### CODE REVIEW

**Comment by JVickery-TBS** [comment]
Just working on some better conditions to prevent multiple jobs queuing up

**Comment by JVickery-TBS** [comment]
@wardi This one is ready for review.

I modified the `IResourceUrlChange` implementation into `IDomainObjectModification` implementation.

The new method to check if a datastore table should be removed (`_should_remove_unsupported_resource_from_datastore`) is quite chunky unfortunately. This was because the entity in the `notify` method is the `ckan.model.Resource` object and not the `dict` like it is in the job callback.

Not sure how to make that slimmer? Would I be able to do something like `res = session.query(model.Resource).get(resource_id)`? A session object does not get passed to the `notify` method, but the class seems to implement some type of session thing: https://github.com/open-data/ckan/blob/canada-v2.8/ckan/model/modification.py

Would I be able to get the session from the toolkit?

File: ckanext/xloader/plugin.py [hunk 1]
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
...
+def _should_remove_unsupported_resource_from_datastore(res_dict_or_obj):
+    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.limit_datastore_tables', False)):
+        return False
+    if isinstance(res_dict_or_obj, Resource):
+        return ((not XLoaderFormats.is_it_an_xloader_format(getattr(res_dict_or_obj, 'format', u''))
+                    or getattr(res_dict_or_obj, 'url_changed', False))
+                and (getattr(res_dict_or_obj, 'url_type') == 'upload'
+                    or getattr(res_dict_or_obj, 'url_type') == '')
+                and (getattr(res_dict_or_obj, 'datastore_active', False) 
+                    or getattr(res_dict_or_obj, 'extras', {}).get('datastore_active', False)))
```
Comment (wardi) [inline ckanext/xloader/plugin.py]: this is kind of a hack, but so is passing either a dict or Resource object to the same function.
```suggestion
        res_dict_or_obj = res_dict_or_obj.__dict__
```

File: ckanext/xloader/plugin.py [hunk 2]
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
...
+    """
+    lc = ckanapi.LocalCKAN()
+    try:
+        res = lc.action.resource_show(id=resource_id)
+        pkg = lc.action.package_show(id=res['package_id'])
+    except toolkit.ObjectNotFound:
+        log.error('Resource %s does not exist.' % res['id'])
+        return
+
+    # only remove datastore tables from dataset types
```
Comment (wardi) [inline ckanext/xloader/plugin.py]:
```suggestion
    # only remove datastore tables from dataset types (canada fork only)
```

File: ckanext/xloader/plugin.py [hunk 3]
```diff
@@ -208,3 +229,50 @@ def get_helpers(self):
             xloader_helpers.xloader_status_description,
             'is_xloader_format': xloader_helpers.is_xloader_format,
         }
+
+
+def _should_remove_unsupported_resource_from_datastore(res_dict_or_obj):
+    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.limit_datastore_tables', False)):
```
Comment (wardi) [inline ckanext/xloader/plugin.py]: A better config option name, remember to document it.

```suggestion
    if not toolkit.asbool(toolkit.config.get('ckanext.xloader.clean_datastore_tables', False)):
```

File: ckanext/xloader/plugin.py [hunk 4]
```diff
@@ -109,11 +113,28 @@ def configure(self, config_):
...
+            return
 
-    def notify(self, resource):
-        # disable automatic submission of resource to xloader if validation is enabled
-        if 'validation' in config.get('ckan.plugins'):
+        if _should_remove_unsupported_resource_from_datastore(entity):
+            p.toolkit.enqueue_job(fn=_remove_unsupported_resource_from_datastore, args=[entity.id])
+
+        # disable automatic submission of resource to xloader
+        # if validation is enabled or the url has not changed
```
Comment (wardi) [inline ckanext/xloader/plugin.py]:
```suggestion
        # if validation is enabled or the url has not changed (canada fork only)
```

**Comment by JVickery-TBS** [comment]
@wardi okay I think this should be good now. The new config option has been added to the readme (will go into the declaration yaml upstream of course too).

I removed the github test workflow for this fork now too.

**Comment by wardi** [comment]
can we add a flag to disable tests instead of removing the whole files, to reduce future conflicts?

**Comment by JVickery-TBS** [comment]
@wardi workflow file brought back and just changed on action to dispatch.

**Review by wardi — APPROVED** [review]

**Pull request merged**